### ISSUE TITLE
Loss function with additional parameters - asymmetric sample weights

### ISSUE THREAD

**Issue by dickreuter** [keras-team/keras#8559]
I'm looking for a way to create a loss function that looks like this:
The function should then maximize for the reward. Is this possible to achieve in Keras?
Any suggestions how this can be achieved are highly appreciated. 

```
def special_loss_function(y_true, y_pred, reward_if_correct, punishment_if_false):
    loss = if binary classification is correct apply reward for that training item
           if binary classification is wrong, apply punishment for that training item
                    )
    return K.mean(loss, axis=-1)
```
The important part is that each individual sample will have a custom weight.

**Comment by mpariente** [comment]
If you want to define a loss in the way you described, the function needs to have only two arguments : y_true and y_pred. 
If you'd define the reward and punishment as constants, you wouldn't have to pass them and I guess it could work.

Otherwise, the common way to create a custom loss function with more than y_true and y_pred is to use Lambda function to compute the loss. You should check [this example](https://github.com/fchollet/keras/blob/master/examples/image_ocr.py)
The loss is defined [here](https://github.com/fchollet/keras/blob/master/examples/image_ocr.py#L331) and used [here](https://github.com/fchollet/keras/blob/master/examples/image_ocr.py#L475)

**Comment by dickreuter** [comment]
Ok thanks for this, I understand that in the example you referenced ctc_batch_cost is used and the loss function cuts the first two iterations of that evaluation as they are not desired. For that it uses a Lambda layer class, to generate a custom loss function. Isn't it a bit counter-intuitive to use a layer function to create a loss function?

The approach I've been looking at for my example is to pass in the weights along with y_true and then cut the tensor into two, separating out the weights and the y_true as seen below. Would an approach like this be possible at all or would this interfere with the normalization process etc?

```
def decompose_y_true(y_true_and_weights):
    y_true = y_true_and_weights[:,1]
    weights= y_true_and_weights[:,1:]
    return y_true, weights


def custom_loss(y_true_and_weights, y_pred):
    y_true, y_weights = decompose_y_true(y_true_and_weights)
     loss = # some loss operation
    return K.mean(loss, axis=-1)

```

Much more elegant would be if I could pass in my weights over the sample_weights parameter in the fit function, but it seems there are some limits what shape those weights can have, and also there's no way to retrieve them within the loss function as far as I can tell. Or is there any way to pass this into the loss function somehow so I can operate on them from there?

**Comment by mpariente** [comment]
In the loss function, `y_true` and `y_pred` should have exactly the same shape, so adding the weights inside of `y_true` won't work..

Do you want to weight the contribution of each item to the final loss or what do you want exactly? I cannot help that much without some details. 
In the general case, I would work with a Lambda layer, which, if you want to weight contributions in the final loss, would make perfect sense.

**Comment by dickreuter** [comment]
Yes, each sample item has an individual contribution to the loss, but in addition to that the contribution also depends on whether the binary_cross_entropy loss function considers y_true to be equal to y_pred or not. Depending on that outcome, a different weight needs to be applied.
